### PR TITLE
James 3779 Tasks are doing sneaky blocking calls

### PR DESCRIPTION
**Pull request by vttranlina**
Jira: https://issues.apache.org/jira/browse/JAMES-3779

### CODE REVIEW

File: server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/service/DeleteMailsFromMailQueueTask.java [hunk 1]
```diff
@@ -177,14 +182,15 @@ Optional<MailAddress> getMaybeSender() {
...
-                    optionalName, optionalRecipient, Clock.systemUTC().instant())));
+    public Publisher<TaskExecutionDetails.AdditionalInformation> detailsReactive() {
+        return Mono.justOrEmpty(lastAdditionalInformation)
+            .switchIfEmpty(Mono.fromCallable(() -> this.queue.map(queue ->
+                    new AdditionalInformation(
+                        queueName,
+                        initialCount.get(),
+                        getRemainingSize(queue), optionalSender,
+                        optionalName, optionalRecipient, Clock.systemUTC().instant())))
+                .handle(publishIfPresent()));
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/service/DeleteMailsFromMailQueueTask.java:193]: .subscribeOn ?

File: server/task/task-memory/src/main/java/org/apache/james/task/TaskManagerWorker.java [hunk 2]
```diff
@@ -32,20 +32,39 @@ interface Listener {
...
 
+        Publisher<Void> failed(TaskId taskId, Publisher<TaskExecutionDetails.AdditionalInformation> additionalInformationPublisher);
+
         Publisher<Void> cancelled(TaskId taskId, Optional<TaskExecutionDetails.AdditionalInformation> additionalInformation);
 
+        Publisher<Void> cancelled(TaskId taskId, Publisher<TaskExecutionDetails.AdditionalInformation> additionalInformationPublisher);
+
         Publisher<Void> updated(TaskId taskId, TaskExecutionDetails.AdditionalInformation additionalInformation);
+
+        Publisher<Void> updated(TaskId taskId, Publisher<TaskExecutionDetails.AdditionalInformation> additionalInformationPublisher);
```
Comment (chibenwa) [inline server/task/task-memory/src/main/java/org/apache/james/task/TaskManagerWorker.java:55]: Do we need to keep the old methods around?

File: server/mailrepository/mailrepository-cassandra/src/main/java/org/apache/james/mailrepository/cassandra/CassandraMailRepository.java [hunk 3]
```diff
@@ -156,6 +157,11 @@ public long size() {
         return countDAO.getCount(url).block();
     }
 
+    @Override
+    public Publisher<Long> sizeAsReactive() {
```
Comment (chibenwa) [inline server/mailrepository/mailrepository-cassandra/src/main/java/org/apache/james/mailrepository/cassandra/CassandraMailRepository.java:161]: sizeReactive IMO

File: server/task/task-memory/src/main/java/org/apache/james/task/TaskManagerWorker.java (hunk 2 above)
Comment (vttranlina) [reply server/task/task-memory/src/main/java/org/apache/james/task/TaskManagerWorker.java]: Removed un-used method

File: server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/service/ClearMailQueueTask.java [hunk 4]
```diff
@@ -110,7 +111,9 @@ public Result run() {
             this.initialCount = Optional.of(getRemainingSize(queue));
             this.queue = Optional.of(queue);
             queue.clear();
-            this.lastAdditionalInformation = details();
+            this.lastAdditionalInformation = Mono.from(detailsReactive())
+                .subscribeOn(ReactorUtils.BLOCKING_CALL_WRAPPER)
+                .block();
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/service/ClearMailQueueTask.java:116]: Why?

File: server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/service/DeleteMailsFromMailQueueTask.java [hunk 5]
```diff
@@ -149,7 +150,9 @@ public Result run() {
             optionalRecipient.ifPresent(Throwing.consumer(
                 (MailAddress recipient) -> queue.remove(ManageableMailQueue.Type.Recipient, recipient.asString())));
 
-            this.lastAdditionalInformation = details();
+            this.lastAdditionalInformation = Mono.from(detailsReactive())
+                .subscribeOn(ReactorUtils.BLOCKING_CALL_WRAPPER)
+                .block();
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/service/DeleteMailsFromMailQueueTask.java:155]: Why?

File: server/task/task-api/src/main/scala/org/apache/james/task/TaskExecutionDetails.scala [hunk 6]
```diff
@@ -15,24 +15,27 @@
...
 
 object TaskExecutionDetails {
 
   trait AdditionalInformation {
     def timestamp: Instant
   }
 
-  def from(task: Task, id: TaskId, hostname: Hostname) = new TaskExecutionDetails(id, task.`type`, WAITING, submittedDate = ZonedDateTime.now, submittedNode = hostname, () => task.details)
+  def from(task: Task, id: TaskId, hostname: Hostname) = new TaskExecutionDetails(id, task.`type`, WAITING, submittedDate = ZonedDateTime.now, submittedNode = hostname,
+    () => SMono.fromPublisher(task.detailsReactive()).subscribeOn(ReactorUtils.BLOCKING_CALL_WRAPPER).block())
```
Comment (chibenwa) [inline server/task/task-api/src/main/scala/org/apache/james/task/TaskExecutionDetails.scala:38]: This .block call is kink of uggly

**Review by chibenwa — COMMENTED** [review]
We should avoid .subscribeOn before .block...

File: server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/service/ClearMailQueueTask.java (hunk 4 above)
Comment (vttranlina) [reply server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/service/ClearMailQueueTask.java:116]: Ah, I should not subscribeOn blocking wrapper. My miss

**Comment by chibenwa** [comment]
Can you please squash fixups?

**Comment by chibenwa** [comment]
Please check [org.apache.james.app.spring.JamesSpringContextTest.springShouldLoadAndAddOnlyOneQuotaUpdaterListener](https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-1049/9/testReport/junit/org.apache.james.app.spring/JamesSpringContextTest/springShouldLoadAndAddOnlyOneQuotaUpdaterListener/)

**Review by Arsnael — APPROVED** [review]

**Comment by vttranlina** [comment]
The error from ci is not related to this pr, Please restart the ci

**Review by quantranhong1999 — APPROVED** [review]

**Comment by vttranlina** [comment]
```
org.apache.james.mpt.imapmailbox.lucenesearch.LuceneFetchBodyStructureTest.testFetchFetchMultipartBodyStructureITALY
Error Message
Location: /org/apache/james/imap/scripts/FetchMultipartBodyStructure.test:209
LastClientMsg: A6 FETCH 1:* ()
Expected: 'A6 OK FETCH completed\.'
Actual   : '* 2 EXISTS'
```
It looks like not related to this pr, I test on local and it passed. 
I will rebase the latest master

**Comment by Arsnael** [comment]
Related or not? https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-1049/36/testReport/junit/org.apache.james.task.eventsourcing.distributed/DistributedTaskManagerTest/badPayloadShouldNotAffectTaskManagerOnCancelTask/

**Comment by vttranlina** [comment]
> Related or not? https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-1049/36/testReport/junit/org.apache.james.task.eventsourcing.distributed/DistributedTaskManagerTest/badPayloadShouldNotAffectTaskManagerOnCancelTask/

Maybe related, But I re-test on local, and it passed. 
Let me more checking

**Comment by vttranlina** [comment]
Tried it many times, but the test passed
image

**Comment by vttranlina** [comment]
```
reactor.core.Exceptions$RetryExhaustedException: Retries exhausted: 3/3
Caused by: com.rabbitmq.client.AlreadyClosedException: connection is already closed due to connection error; protocol method: #method<connection.close>(reply-code=320, reply-text=CONNECTION_FORCED - broker forced connection closure with reason 'shutdown', class-id=0, method-id=0)
```
it's not related to this pr, it will be resolved at https://github.com/apache/james-project/pull/1075

**Pull request merged**